AIP Extensions (#520)
googleapis/aip: aip/0008.md

link to the relevant AIP. AIP links **may** point to a particular section of
the AIP if appropriate.

### Static AIP Reference

Most times, you will likely want to use a Context Based Reference when
providing a link to an AIP (see below). However it's worth noting that there
may be times when you want to refer to a specific AIP from a specific context,
regardless of what context you are in.

For example, an organization based AIP may extend AIP-121 by stating that
everything in the parent AIP-121 is valid, and that the context based AIP-121
only provides some additional requirements. In this case, the organization
level AIP-121 would want to provide a link to the domain level AIP-121. If the
link were a Context Base Reference, the AIP link to AIP-121 would just link

> back to the organization level AIP!

Reviewer: If we use fork and merge, presumably the content will just be duplicated.

Author: I made up some language here, so let me clarify. I'm opening to using different terms, or maybe clarifying as a team that we don't need these terms.

![Screen Shot 2020-06-10 at 10 45 34 AM](https://user-images.githubusercontent.com/38894273/84295385-bea7a500-ab07-11ea-81f5-eed44176c808.png)

I don't think the current nav menu supports this very well. As mentioned before, I'm not super happy with the nav menu, and want to revisit it once we better define what all this looks like.

I perceive the end result eventually looking like the image above.

Domain level AIPs would be the AIPs that belong to the domain, i.e. google.aip.dev, or salesforce.aip.dev. Although salesforce.aip.dev may have forked from google.aip.dev, it does not guarantee that it would have the same level AIPs. All domain level AIPs should not have an prefix (i.e. `ACU-XXXX`, unless they do not exist in the AIP spec (whether that spec is implicit based on our general AIP definition or explicit via an actual schema).

Org level AIPs are the extensions on the domain level. All of these would have a prefix. Rather than duplicating the domain level AIPs within the org, they would only define their differences with the domain AIPs. When referring to an AIP that has not been altered in the org, the AIP site will just display the domain level AIP.

This allows for multiple org extensions to exist in the same repository/site as the domain level AIPs, which is what we are looking for in Salesforce. Interested in understanding if our visions are unaligned or conflict.

The section below discusses context based references in more detail, and
provides examples of how these links should be implemented. This section is
provided for context for these examples, as the implementation examples cannot

> really be shared without consider both types of references together.

Reviewer 2: ```suggestion
really be shared without considering both types of references together.
```

### Context Based References

It is not uncommon for AIPs to have relationships or even dependencies on one
another. For example, it may be logical for some implementations of AIP-131
(Methods: GET) to reference AIP-121 (Resource Names), as the address used by
the method should follow the guidelines for resource naming. It may also be
possible for each of these AIPs to be overridden at an organizational level,
yet still maintain their relationship one with another.

In such a case, it would be important that the links between these two AIPs
took into consideration that context under which they were being viewed. For
example if AIP-131 was being viewed under the organization context, then all
links to AIP-121 should point to the organizational extension of AIP-121, and
not the domain level implementation of AIP-121. Likewise, if AIP-121 is being
viewed at the domain level, then links to AIP-131 should point to the domain
level implementation of AIP-131, and not to any organizational extensions.

> All AIP reference links that need to take context into consideration **must**

Reviewer: Do we need to do literally anything here? Seems like fork and merge plus static site generation will just handle it since we use `./0100.md` style links.

</script>
<a id="{{ aip }}">AIP-{{ aip }}</a>
```

Pros:

- Pretty clean implementation from the client's perspective
- Implementing a static reference would be straight forward

Cons

- utilizes includes, which is slow

#### Automagic

Automatically looks at local refs to see if they exist.

**What the AIP Editor Adds**

```
./0131.md
```

**What the solution looks like** Javascript file that is added to each aip
file:

```
$.when($.ready)(() => {
  const context = localStorage.get("context")

  const anchors = window.getElementById("aip-main").getElementsByTagName("a")
  let href
  for(i = 0; i anchors.length; i++) {
    href = anchors.getAttribute("href")

    if( href is relative) {
      if({{ file_exits("/"context"/"href) }})
        use /context/href
      else
        use ./href
    }
  }
})
```

Pros:

- client does not have to do anything for context based references

Cons:

- client will need to do something different for static references
- javascript iterates over all the anchors in the aip-main element, and
  potentially alters them. This is just more overhead when loading a page. Not
  huge, but still there.
- feels a little hackish.

#### Class Based

> Takes advantage of a kramdown feature for providing css classes to links

Reviewer 2: ```suggestion
Takes advantage of a [kramdown](https://kramdown.gettalong.org/) feature for providing css classes to links
```
Without a link, it can be considered a typo.

inline.

**What the AIP Editor Adds** Must be done when defining the reference:

```
[This][]{.context} is a context based reference.
[This][]{.static} is a static reference.
```

**What the implementation looks like** implementation mostly looks like the
automagical one, except we are looking for this class.

```
$.when($.ready)(() => {
  const context = localStorage.get("context")

  const anchors = window.getElementById("aip-main").getElementsByClass("context")
  let href
  for(i = 0; i anchors.length; i++) {
    href = anchors.getAttribute("href")

    if( href is relative) {
      if({{ file_exits("/"context"/"href) }})
        use /context/href
      else
        use ./href
    }
  }
})
```

Pros:

- declarative, similar solution for both static and context based references
- optional, although the aip editor **should** use the declarative solution, if
  they don't we can still treat the default as `static`.
- feels less hackish as it takes advantage of existing functionality.

Cons:

- tighter coupling with kramdown.

> #### Failover Redirects

Reviewer: This is clever and I did not think of it. :-)

Worth mentioning, but I have not gone too deep into it as I have the above
solutions. We can also write our own `redirect.html` file in the `_layouts`

> directory and may be able to add some logic to fail over to a parent AIP if the

Reviewer 2: ```suggestion
directory and may be able to add some logic to failover to a parent AIP if the
```

context based one doesn't work.

Pros:

- Way to determine whether an AIP reference is context based or static may be
  allowed to be more arbitrary and easier for aip editor to implement.

Cons:

- redirects in GitHub pages can only be done via meta-refresh. If a context
  mostly inherits from it's parent except for one or two files, that means that
  the request process for most context based AIP references would be as
  follows:

> 1. Request non existent page

Reviewer 2: ```suggestion
1. Request non-existent page
```

2. Jekyll redirects to our custom redirect page
3. redirect page is returned to the browser, client should not see if it fall
   goes well.
4. redirect page creates new request for parent page.

- The above process could be a pain if there are multiple parents (i.e. an
  extension of an extension).

> 

Author: Hopefully it's clear that not all of this would actually go into the AIP.

The Static AIP Reference and the Context Based Reference sections kind of describe the problem I am trying to solve, and the h4 sections below that describe some of the solutions I have considered. This is where I am looking for feedback.
